lib/controllers/book_controller.dart

```diff
@@ -2,55 +2,205 @@ import 'dart:async';
 
 import 'package:flutter_riverpod/flutter_riverpod.dart';
 import 'package:paourtest/data/domain/books_model.dart';
+import 'package:paourtest/data/domain/chapter_model.dart';
+import 'package:paourtest/data/domain/exam_list.dart';
+import 'package:paourtest/data/service/lesson_service.dart';
 import 'package:paourtest/data/usecases/lesson_usecase.dart';
 
 class LessonState {
   final bool loading;
 
   final List<BookModel> books;
+  final List<ExamList> examChapterList;
+  final List<String> subjects;
+  final List<String> levels;
 
   const LessonState({
     required this.loading,
     required this.books,
+    required this.subjects,
+    required this.levels,
+    required this.examChapterList,
   });
 
   LessonState copyWith({
     bool? loading,
     List<BookModel>? books,
+    List<String>? subjects,
+    List<String>? levels,
+    List<ExamList>? examChapterList,
   }) {
     return LessonState(
       loading: loading ?? this.loading,
       books: books ?? this.books,
+      subjects: subjects ?? this.subjects,
+      levels: levels ?? this.levels,
+      examChapterList: examChapterList ?? this.examChapterList,
     );
   }
 }
 
 class LessonController extends StateNotifier<LessonState> {
-  LessonController(this._lessonUseCase) : super(const LessonState(loading: false, books: []));
+  LessonController(this._lessonUseCase)
+      : super(const LessonState(loading: false, books: [], subjects: [], levels: [], examChapterList: []));
   final LessonUseCase _lessonUseCase;
 
   Future<void> fetchInitialBooks() async {
-    print("fetch initial books");
-
     state = state.copyWith(loading: true);
 
     List<BookModel> bookResult = await _lessonUseCase.initBooks();
+
+    bookResult.removeWhere((book) => book.displayTitle == null);
+    List<String> subjects = [];
     for (var book in bookResult) {
-      print(book.url);
+      for (var subject in book.subjects) {
+        if (!subjects.contains(subject)) {
+          subjects.add(subject);
+        }
+      }
     }
-    bookResult.removeWhere((book) => book.displayTitle == null);
+    List<String> levels = [];
+    for (var book in bookResult) {
+      for (var level in book.levels) {
+        if (!levels.contains(level)) {
+          print(level);
+          levels.add(level);
+        }
+      }
+    }
+
+    sort(bookResult);
 
     state = state.copyWith(
       loading: false,
       books: bookResult,
+      subjects: subjects,
+      levels: levels,
     );
   }
 
+  void sort(List<BookModel> bookResult) {
+    sortBySubject(bookResult);
+    sortByLevel(bookResult);
+  }
+
+  void sortByLevel(List<BookModel> bookResult) {
+    List<String> orderedLevels = [
```

Comment:
This is a great idea, and it shows you took time to understand the data intimately.